Cleanup of REST API

letmaik opened the pull request:
It's not clear yet whether and in what form a REST API will be standardized but this PR cleans up the existing section and makes the API vendor-neutral (previously, some CCF-specific bits had leaked into it) including allowing for registration of claims to be a long-running operation which provides more flexibility for implementations.

letmaik commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -552,64 +575,126 @@ Editor's Note: This may be moved to appendix.
+- Status 202 - Registration is successful.
+  - Header `Location: <Base URL>/operations/<Operation ID>`
+  - Header `Content-Type: application/json`
+  - Body `{ "operationId": "<Operation ID>", "status": "registered", "entryId": "<Entry ID"> }`
```
May be worth shortening `"operationId"` to `"id"` since it's obvious from context.

letmaik replied:
I'll leave it as-is for now to keep it clear.

letmaik commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -552,64 +575,126 @@ Editor's Note: This may be moved to appendix.
+- Status 200 - Registration was successful
+    - Header: `Location: <Base URL>/entries/<Entry ID>`
+    - Header: `Content-Type: application/json`
+    - Body: `{ "operationId": "<Operation ID>", "status": "registered", "entryId": "<Entry ID>" }`
```
Status codes should probably be application-agnostic, like `running`, `succeeded`, and `failed`.

letmaik replied:
Changed `registered` to `succeeded`.

letmaik commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -552,64 +575,126 @@ Editor's Note: This may be moved to appendix.
-  - Error code `TransactionInvalid`
+- Status 200.
+  - Header `Content-Type: application/json`
+  - Body `{ "entryId": "<Entry ID>", "status": "registered" }`
```
Probably `"status"` should be removed from this response since any other case would return a 404 anyway.

rjb4standards replied:
NIST labeling requires a date. I suggest replacing status with EffectiveDate.

letmaik replied:
I removed `"status"` without adding more fields for now. Registration date is part of the receipt already, but we can discuss in separate issues what kind of metadata this entry resource should have.

thomas-fossati commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -544,6 +544,29 @@ Editor's Note: This may be moved to appendix.
 
+All messages are sent as HTTP GET or POST requests.
+
+All errors are returned with HTTP 4xx or 5xx status code as JSON objects following the OData JSON (ISO/IEC 20802-2:2016) schema below:
+
+~~~
+{
+  "error": {
+    "code": "<code>",
+    "message": "<message>"
+  }
+}
+~~~
```
Question: why defining a custom format when you can use the standardised "Problem Details" envelope, which is available for both [CoAP](https://datatracker.ietf.org/doc/rfc9290/) and [HTTP](https://datatracker.ietf.org/doc/draft-ietf-httpapi-rfc7807bis/) ?

letmaik replied:
I'm not sure why you say custom. The format I used is standardised in ISO/IEC 20802-2:2016, see the sentence above.

thomas-fossati replied:
it's preferable to depend on open standards rather than standards behind paywalls

letmaik replied:
Sure, I agree. The OData JSON Format is openly standardised under OASIS, see https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#_Toc38457793, and has been republished in ISO, behind a paywall. I proposed this because it's simple (e.g., does not require minting URIs for error codes) and is already widely in use (e.g., all of Microsoft is using it). Happy to continue the discussion though! :)

thomas-fossati replied:
cool, maybe change the ref to OASIS.

FWIW, RFC7807 too is quite widely used (see https://github.com/search?q=RFC+7807), also IETF dogfooding and whatnot :-)

letmaik replied:
Alright, in combination with https://datatracker.ietf.org/doc/rfc7807/referencedby/ I think it's worth following the lead of CT 2.0 https://www.rfc-editor.org/rfc/rfc9162.html#name-log-client-messages and use urns for error types:
```json
    {
        "type": "urn:ietf:params:scitt:error:invalidInput",
        "detail": "..."
    }
```

thomas-fossati replied:
👍 awesome

letmaik replied:
Done, please check.

thomas-fossati replied:
thanks

thomas-fossati commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -552,64 +575,126 @@ Editor's Note: This may be moved to appendix.
-- HTTP Status `201` - Registration was tentatively successful pending service consensus.
-- HTTP Status `400` - Registration was unsuccessful.
-  - Error code `AwaitingDIDResolutionTryLater`
+- Status 202 - Registration is successful.
```
why not 200 here?  Isn't "registered" a final state?

letmaik replied:
The idea is that every submission is uniformly treated as a long-running operation from an API standpoint, even if the operation finishes immediately. This makes clients a little easier to write. Returning the operation object makes really only sense with 202. For 200/201 I would expect either a redirect to the created object or the response being the actual created object. Doing that, clients would have to support both cases.

thomas-fossati replied:
I am not sure I understand the simplicity argument, but regardless that doesn't really match HTTP semantics (see [§15.3.3 of STD97](https://www.rfc-editor.org/rfc/rfc9110.html#name-202-accepted)):

_"[...] the request has been accepted for processing, but the processing has not been completed."_

letmaik replied:
With simplicity, I mean this: https://github.com/scitt-community/scitt-api-emulator/blob/4a56fad2171a4a0fd32075a5378d21b3aa66aab4/scitt_emulator/client.py#L78-L92. If you have two possible answer types (one creating and returning an operation, the other not) then the code to handle that gets more complicated. I found some reference for this here: https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#132-stepwise-long-running-operations "For an API that's defined as a Stepwise Long Running Operation the service MUST go through the Stepwise Long Running Operation flow even if the operation can be completed immediately."
If we have two separate responses, 202 for long-running, and 201 for immediate, which would you return in the 201 case?

thomas-fossati replied:
my preference would be:

### Request

```http
POST /entries HTTP/1.1
Host: scitt.example
Accept: application/json
Content-Type: application/cose

[ COSE Sign1 claims ]
```

### Response

#### Immediate

```http
HTTP/1.1 201 Created
Location: /entries/{entry-id}
Content-Type: application/json

{ "entry-representation": "..." }
```

#### Pending

```http
HTTP/1.1 202 Accepted
Location: {operation-monitor-url}
Content-Type: application/json

{ "operation-monitor-resource-representation": "..." }
```

letmaik replied:
Alright, let's do that. I prototyped it in the emulator and didn't see any issue. There is a little more code for clients but it is minimal and we expect this to be hidden in libraries anyway.

thomas-fossati commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -552,64 +575,126 @@ Editor's Note: This may be moved to appendix.
+One of the following:
 
-`InvalidInput` means either the Signed Claims message is syntactically malformed, violates the signing profile (e.g. signing algorithm), or has an invalid signature relative to the currently resolved DID document.
+- Status 200 - Registration is pending
```
shouldn't this be 202?

letmaik replied:
The operation resource itself is not pending, it just exists. Similarly, you wouldn't return 400 if the operation failed, since the operation resource itself has been returned just fine.

thomas-fossati replied:
oh yes, you are right, sorry for the noise :-)

thomas-fossati commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -552,64 +575,126 @@ Editor's Note: This may be moved to appendix.
+One of the following:
+
+- Status 200.
+  - Header: `Content-Type: application/cose`
```
STD96 defines `application/cose; cose-type="cose-sign1"` for this

letmaik replied:
We required tagged COSE, so I'd rather avoid repeating it here for the sake of simpler clients.

thomas-fossati replied:
sure, but these are different layers.

unless you are only ever allowing COSE Sign1 in the body, in which case the type qualifier is redundant, this approach is prone to MIME sniffing, which in the long term makes clients more complex rather than simpler.

letmaik replied:
My main worry with using a media type with parameter is that it's not trivial to match it. Some clients will add a whitespace, some won't, and the server has to decode the media type and parameters, instead of doing a simple string match.

OR13 replied:
If we can avoid normative requirements for media type parameters, we should... not sure if thats possible in this case.

henkbirkholz replied:
@letmaik Are you sure that we will stick with "COSE Sign1 only" forever?

letmaik replied:
@henkbirkholz No, but I think it'll be more likely that we use more concrete media types in the future, rather than the generic `application/cose;type=..`, like `application/scitt-claim+cose`. So, I'd prefer to treat this (`application/cose`) like for receipts (`application/cbor`) a valid placeholder.

thomas-fossati replied:
> @henkbirkholz No, but I think it'll be more likely that we use more concrete media types in the future, rather than the generic `application/cose;type=..`, like `application/scitt-claim+cose`. So, I'd prefer to treat this (`application/cose`) like for receipts (`application/cbor`) a valid placeholder.

If the plan is to move to more precise typing info (i.e., `application/scitt-claim+cose`) in the near future, then I'm happy.

(Note though that `+cose` is not a registered ["structured syntax suffix"](https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml), so you'd need to either register it or fall back to `+cbor`.)

thomas-fossati commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -552,64 +575,126 @@ Editor's Note: This may be moved to appendix.
+One of the following:
+
+- Status 200.
+  - Header: `Content-Type: application/cbor`
```
is there any plan to add a media type for SCITT Receipts?

letmaik replied:
I assume so. Consider it a valid placeholder.

OR13 reviewed (approved):
Excellent set of changes.

thomas-fossati commented on draft-birkholz-scitt-architecture.md:
```diff
@@ -544,6 +546,31 @@ Editor's Note: This may be moved to appendix.
+
+~~~
+{
+  "type": "urn:ietf:params:scitt:error:badSignatureAlgorithm"
```
```suggestion
  "type": "urn:ietf:params:scitt:error:badSignatureAlgorithm",
```

thomas-fossati replied:
BTW, I _think_ that if you tag the code block with the right "language" (e.g., json, http-message), the build pipeline can run the validators for you.

thomas-fossati replied:
I made a test: it looks like json validation is done with no extra effort but tagging:

```
~~~json
{}
~~~
```

whilst for http validation you need to tag:


```
~~~http-message
POST /stuff HTTP/1.1
Host: stuff.example
Accept: application/stuff
Content-Type: application/cbor

[ CBOR-encoded things ]
~~~
```

as well as adding the http-lint tool as a lint precondition to the Makefile:

```
lint:: http-lint

rfc-http-validate ?= rfc-http-validate
.SECONDARY: $(drafts_xml)
.PHONY: http-lint
http-lint: $(drafts_xml) http-lint-install
	$(rfc-http-validate) $(filter-out http-lint-install,$^)

.PHONY: http-lint-install
http-lint-install:
	@hash $(rfc-http-validate) 2>/dev/null || pip3 install rfc-http-validate typing_extensions
```

letmaik commented:
I addressed all feedback that came in and will move this PR over to the repo in ietf-wg-scitt. More review can happen there. I accidently opened this PR here in the old repo.

thomas-fossati commented:
thanks @letmaik

letmaik commented:
Now here: https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/pull/3
